### PR TITLE
Add support for Python 3.12 wheel builds

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** This PR updates the maximum tested version of Python to 3.12, and enables build support for wheels with said version.

**Description of the Change:** Add support for Python 3.12 builds and testing. Upgrade pybind11. Upgrade cibuildwheel.

**Benefits:** Adds support for Python 3.12 wheels.

**Possible Drawbacks:** Additional effort during release time due to extra Python versions.

**Related GitHub Issues:**

### CODE REVIEW

**Review by multiphaseCFD — APPROVED** [review]
Looks good to me! Thanks @mlxd !

File: .github/workflows/set_wheel_build_matrix.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: Set wheel build matrix
 
 env:
   PYTHON3_MIN_VERSION: "9"
-  PYTHON3_MAX_VERSION: "11"
+  PYTHON3_MAX_VERSION: "12"
```
Comment (maliasadi) [inline .github/workflows/set_wheel_build_matrix.yml:5]: Do you want to update the changelog?

**Review by maliasadi — APPROVED** [review]
:tada:

File: .github/workflows/set_wheel_build_matrix.yml (hunk 1 above)
Comment (mlxd) [reply .github/workflows/set_wheel_build_matrix.yml:5]: Yep, will do that prior to merge if all else looks good.

**Pull request merged**